File: 프로그래머스/2/142085. 디펜스 게임/디펜스 게임.cpp
#include <bits/stdc++.h>
using namespace std;


int solution(int n, int k, vector<int> enemy) {
    int answer = 0;

    priority_queue<int> pq;
    int rounds = 0;

    for(int i =0; i<enemy.size(); i++){

        n -= enemy[i];
        pq.push(enemy[i]);

        if(n < 0){
            if(k > 0){
                n += pq.top();
                pq.pop();
                k--;
            }else{
                break;
            }
        }
        answer = i + 1;
    }
    return answer;
}

File: 프로그래머스/2/142085. 디펜스 게임/디펜스 게임.java
import java.util.PriorityQueue;
class Solution {
  public int solution(int n, int k, int[] enemy) {
    int answer = 0;

    PriorityQueue<Integer> q = new PriorityQueue<>();
    
    for(int i =0 ; i<enemy.length; i++){
      int e = enemy[i];
      q.add(-e);
      n -= e;
      
      if(n < 0){
        if(k > 0){
          k--;
          n-= q.poll();
        }else{
          break;
        }
      }
      
      answer = i + 1;
    }
    
    return answer;
  }
}
